tx_validator.go

```diff
@@ -47,3 +69,156 @@ func (v *WitnessValidatorV0) ValidateWitnesses(newAsset *asset.Asset,
+	chainBridge tapgarden.ChainBridge
+
+	assetStore *tapdb.AssetStore
+
+	proofFile *proof.File
+}
+
+// NewProofChainLookup creates a new ProofChainLookup instance.
+func NewProofChainLookup(chainBridge tapgarden.ChainBridge,
+	assetStore *tapdb.AssetStore, proofFile *proof.File) *ProofChainLookup {
+
+	return &ProofChainLookup{
+		chainBridge: chainBridge,
+		assetStore:  assetStore,
+		proofFile:   proofFile,
+	}
+}
+
+// CurrentHeight return the current height of the main chain.
+func (l *ProofChainLookup) CurrentHeight(ctx context.Context) (uint32, error) {
+	return l.chainBridge.CurrentHeight(ctx)
+}
+
+// TxBlockHeight returns the block height that the given transaction was
+// included in.
+func (l *ProofChainLookup) TxBlockHeight(ctx context.Context,
+	txid chainhash.Hash) (uint32, error) {
+
+	// If we don't have a proof available as context, we can only look up
+	// the transaction in the database. Querying it on-chain would cause a
+	// re-scan which might be very time costly for light clients.
+	if l.proofFile == nil || l.proofFile.NumProofs() == 0 {
+		return l.assetStore.TxHeight(ctx, txid)
+	}
+
+	// Let's walk back the proof chain and try to find the transaction.
+	height, err := findTxHeightInProofFile(l.proofFile, txid)
+	switch {
+	case errors.Is(err, errTxNotFound):
+		// Our last ditch attempt is to look up the transaction in the
+		// database. But we might not have it there if the proof is for
+		// a transaction that happened before the asset reached our
+		// node.
+		return l.assetStore.TxHeight(ctx, txid)
+
+	case err != nil:
+		return 0, fmt.Errorf("error fetching proof from context file: "+
+			"%w", err)
+	}
+
+	return height, nil
+}
+
+// findTxHeightInProofFile is a helper function that recursively searches for
+// the block height of a transaction in a proof file.
+func findTxHeightInProofFile(f *proof.File, txid chainhash.Hash) (uint32,
+	error) {
+
+	for i := f.NumProofs() - 1; i >= 0; i-- {
+		p, err := f.ProofAt(uint32(i))
+		if err != nil {
+			return 0, fmt.Errorf("error fetching proof from "+
+				"file: %w", err)
+		}
+
+		if p.AnchorTx.TxHash() == txid {
+			return p.BlockHeight, nil
+		}
+
+		for idx := range p.AdditionalInputs {
+			additionalInput := p.AdditionalInputs[idx]
+			height, err := findTxHeightInProofFile(
+				&additionalInput, txid,
+			)
+			switch {
+			case errors.Is(err, errTxNotFound):
+				continue
+
+			case err != nil:
+				return 0, fmt.Errorf("error fetching proof "+
+					"from additional input file: %w", err)
+			}
+
+			return height, nil
+		}
+	}
+
+	// If we arrive here, we couldn't find the transaction in the proof
+	// file.
+	return 0, errTxNotFound
+}
+
+// MeanBlockTimestamp returns the timestamp of the block at the given height as
+// a Unix timestamp in seconds, taking into account the mean time elapsed over
+// the previous 10 blocks.
+func (l *ProofChainLookup) MeanBlockTimestamp(ctx context.Context,
+	blockHeight uint32) (time.Time, error) {
```

Comment:
Yes, I lifted it. Because otherwise I would've needed to implement the `HeaderCtx` interface which would probably have resulted in way more lines of codes (and awkward either empty methods because they're not used by that function or "proper" implementations that don't serve a real purpose).
So just lifting the logic instead seemed way more straight-forward.